### PR TITLE
KAFKA-4607: Validate the names of auto-generated internal topics

### PR DESCRIPTION
**Pull request by nixsticks**
I considered catching errors to add further information about naming internal state stores. However, Topic.validate() will throw an error that prints the offending name, so I decided not to add too much complexity.

### CODE REVIEW

**Comment by mjsax** [comment]
I think it is better to add this check closer to the user to avoid deep stack traces. Thus, I would add it to `KGroupedStreamImpl` and `KGroupedKTabelImpl`.

\cc @guozhangwang @dguy @enothereska

**Comment by mjsax** [comment]
Furthermore, can you please add a test for this. And please update the corresponding JavaDocs and state that store names but be valid topic names (and explain what this means, ie, which characters are allowed and which not). Thx.

**Comment by nixsticks** [comment]
Will look into it. Thanks @mjsax !

**Comment by nixsticks** [comment]
@mjsax Do you think it might be good to put it both in `KGroupedStreamImpl`/`KGroupedTableImpl` and also in `InternalTopicConfig` in case other, new code tries to use InternalTopicConfig as well?

**Comment by mjsax** [comment]
Sounds like a good idea to me. Maybe we can put it as an assertion into `InternalTopicConfig` ?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (nixsticks) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java:299]: I did a couple of cleanups like this in the cases where it seemed like the test name did not match the actual test, since I am technically writing and editing tests regarding the state store names -- see also `shouldNotAcceptNullStateStoreSupplierWhenAggregatingSessionWindows` and `shouldNotAcceptNullStateStoreSupplierWhenCountingSessionWindows`

**Comment by nixsticks** [comment]
@mjsax There are now ... a lot of changes. I am fairly new to contributing to open source so please let me know if I have updated the pull request in the wrong way. I left the full validation in `InternalTopicConfig` since I figured that internal topics should also follow the rules of regular topics.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (nixsticks) [inline clients/src/main/java/org/apache/kafka/common/Topic.java]: This was taken from the Scala version of this utility class and I am not sure what best to call it

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (mjsax) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java:299]: The basic rule is to have a single PR for a single JIRA -- changes like this add "noise" and complicate reviewing in general (review need to know/judge what is part of the actual PR and what is cleanup). For this case it is ok though IMHO :)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: Please add this to `StatsStoreSupplier#name()`, too? (plus add a test for it)

File: streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KGroupedStream.java]: Nit: we try to follow the following policy to make reviewing easier:
 - each new sentence start a new line (you do follow this already)
 - no line longer then 120 character (otherwise lines are longer then Github displays and you need to scroll to the right)

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: This adds code redundancy. Please remove. Same blow. Only add validation to the already present not-null checks.

Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImpl.java]: Remove: same as above.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImpl.java]: Again. Remove.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java]: Use a class member and use in all test methods to avoid code duplication.
Please add `final` keyword wherever possible.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImplTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImplTest.java]: Use class final class member.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: That's a tricky one... we do not want to have code duplication -- but we also do not want to depend on `kafka-core` module...
\cc @guozhangwang What do you think?

**Review by mjsax — CHANGES_REQUESTED** [review]
Thanks for the update. I have some comment. Overall look already quite good. Fee free to wait with an update until @guozhangwang replied.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImplTest.java
Comment (nixsticks) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedTableImplTest.java]: 👍

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java
Comment (nixsticks) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KGroupedStreamImplTest.java]: 👍

**Comment by miguno** [comment]
What's the status of this PR?  I'd be great to get it merged soon.

**Comment by guozhangwang** [comment]
Seems @nixsticks has not addressed the comments yet.

**Review by guozhangwang — COMMENTED** [review]
A meta comment: instead of checking validity in the builder, could we just check validity once in the `InternalTopicManager`, before calling `StreamsKafkaClient` to create them, just like what we did in the scala's `AdminUtils` pattern?

The javadocs addition looks good to me.

**Comment by mjsax** [comment]
@guozhangwang I prefer shallow stack traces and thus IMHO we should check as soon as we can if names are valid. (Of course, just my personal opinion.)

**Comment by guozhangwang** [comment]
> @guozhangwang I prefer shallow stack traces and thus IMHO we should check as soon as we can if names are valid. (Of course, just my personal opinion.)

I'm convinced.

**Comment by nixsticks** [comment]
@guozhangwang Sorry, I was going to wait for an update on the kafka-core module vs code duplication, and then I went on PTO for two weeks! I will try to get my changes in soon (I'm returning to the US in a few days), but before that, would you mind addressing the comment from @mjsax about the above?

**Comment by nixsticks** [comment]
@guozhangwang @mjsax Decided to open my computer just for a few minutes and then ended up making most of the changes you mentioned. I have not rebased with the upstream yet but will do so after  Could you let me know whether the addition of `Topic.java` is acceptable or whether I should do something else (and whether the changes look acceptable as well)? Thank you!

**Comment by nixsticks** [comment]
Not sure what happened to my previous comment, but: I ended up addressing most of the comments here tonight. Could you let me know whether the changes are acceptable, and what you'd like regarding the addition of `Topic.java` vs using the kafka-core module? Thanks!

**Review by mjsax — APPROVED** [review]
LGTM.

I think we cannot avoid code duplication (cf. newly added `Topic.java` that got copied from core's Scala version) because a dependency to core module is a no-go. And I have not better idea to resolve the issues.

**Comment by mjsax** [comment]
\cc anyof @dguy @enothereska for second review

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (enothereska) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: cc @ijuma

**Comment by enothereska** [comment]
LGTM, but I'd like @ijuma to comment on adding Topic.java please.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (dguy) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: Should we change the scala version to use this one in `common`? i believe `kafka-core` depends on `common`. It would be better to not have the code duplicated if we can avoid @ijuma

**Review by dguy — COMMENTED** [review]
Overall LGTM, but ideally we'd not duplicate the `Topic` code.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: Yes, if we move the logic here, then we should reuse it and not duplicate it (including tests if they exist). A couple more things:

1. This class should be in `common.internals` since we it's an internal class.
2. We are doing this validation client-side, which could potentially be different from the broker validation (e.g. a future version could allow longer topic names or restrict it further due to file system or OS restrictions). It would be good to explain what is the current behaviour to make it clear how this improves on that.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/Topic.java]: This should be saved in a static final field instead of compiling the pattern on every invocation.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/Topic.java]: Nit: `topic.isEmpty`.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/Topic.java]: Nit: let's capitalise all error messages.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/Topic.java]: Nit: if we use a single quote, we avoid the need for the ugly escaping.

**Comment by guozhangwang** [comment]
ping @nixsticks in case you did not receive notifications on new comments.

**Comment by nixsticks** [comment]
@guozhangwang Thanks. Apart from minor comments I'd like to hear more of a solid decision on where this code should go before I make changes again.

**Comment by guozhangwang** [comment]
It is not ideal, though I feel we cannot avoid code duplication on Scala / Java, but I think it's better to go into o.a.k.common.internals instead of o.a.k.common since the latter is a public package meaning that `Topic` will be exposed to users and it'll be hard to change its interfaces in the future. cc @ijuma

**Comment by nixsticks** [comment]
@guozhangwang Thank you, sounds good. I'm a little pressed for time at work right now, so will look into this on the weekend. 👍

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (nixsticks) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: I believe in Java you can only use single quotes for chars, not strings?

Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: @nixsticks That's not what @ijuma meant -> `InvalidTopicException("topic name cannot be '.' or '..'")`
However, I agree that double quote are better, because using single quotes would not be sound, as it checks for equality -- not containment of single chars.

Btw: It's the same exception text as in `kafka.common.Topic`.

**Comment by nixsticks** [comment]
@guozhangwang Addressed minor comments and moved `Topic` into o.a.k.common.internals, along with its test.

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (nixsticks) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: @mjsax Ah I see, thanks! And yes, I did take it from the existing Scala `Topic` class.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: @mjsax I don't understand your soundness comment. The quotation in this case is simply used as an English language device and both styles are accepted (it seems that single quotes are preferred in British English, but double quotes are preferred in American English). Anyway, it's a minor point.

**Comment by guozhangwang** [comment]
@nixsticks There is a checkstyle failure from jenkins builds:

```
:clients:checkstyleMain[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk7-scala2.10/clients/src/main/java/org/apache/kafka/common/internals/Topic.java:1: Line does not match expected header line of '/*'. [Header]
 FAILED
```

File: clients/src/main/java/org/apache/kafka/common/Topic.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/common/Topic.java]: @mjsax could you reply on that?

**Comment by guozhangwang** [comment]
@nixsticks please see https://github.com/apache/kafka/pull/2679 as a reference for the apache headers checkstyle issue.

**Comment by nixsticks** [comment]
Thanks @guozhangwang, fixed the checkstyle error.

File: clients/src/main/java/org/apache/kafka/common/internals/Topic.java
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/common/internals/Topic.java]: Nit: can we add `{` and `}` for all blocks.

File: clients/src/main/java/org/apache/kafka/common/internals/Topic.java
Comment (mjsax) [inline clients/src/main/java/org/apache/kafka/common/internals/Topic.java]: can be package private -- no need to be public (same for `containsOnlyPeriods`, `exceedsMaxLength` and `containsInvalidCharacters`)

File: clients/src/test/java/org/apache/kafka/common/internals/TopicTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/common/internals/TopicTest.java]: this four lines can just be one line: `Topics.validate(topicName)`

File: clients/src/test/java/org/apache/kafka/common/internals/TopicTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/common/internals/TopicTest.java:74]: We should add a test for `containsOnlyPeriods`.

Also `Topics.validate` should be tested for invalid topics -- ie, reverse testto `shouldRecognizeValidTopicNames`

**Review by mjsax — CHANGES_REQUESTED** [review]
We should improve tests coverage. Overall this is in good shape.

File: clients/src/main/java/org/apache/kafka/common/internals/Topic.java
Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/common/internals/Topic.java]: We don't require braces in such cases in our Java code. I think that has been done in the Streams module, but it's not something that has been agreed generally.

Comment (mjsax) [reply clients/src/main/java/org/apache/kafka/common/internals/Topic.java]: Ok. Fair enough. My comment can be ignored than.

**Comment by nixsticks** [comment]
@mjsax Addressed comments.

File: clients/src/test/java/org/apache/kafka/common/internals/TopicTest.java
Comment (mjsax) [inline clients/src/test/java/org/apache/kafka/common/internals/TopicTest.java]: Nit: rename to `shouldThrowOnInvalidTopicNames`

Comment (nixsticks) [reply clients/src/test/java/org/apache/kafka/common/internals/TopicTest.java]: OK, changed.

**Comment by mjsax** [comment]
@nixsticks There was a checkstyle error in the build. Can you have a look. Thx.

**Comment by nixsticks** [comment]
@mjsax I did, last night, but it is for a file (ByteArrayConverter on Connect) that I did not touch.

`checkstyleMain[ant:checkstyle] [ERROR] /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk8-scala2.11/connect/runtime/src/main/java/org/apache/kafka/connect/converters/ByteArrayConverter.java:1: Line does not match expected header line of '/*'. [Header]`

**Comment by mjsax** [comment]
Try to rebase to `trunk`

**Comment by guozhangwang** [comment]
> @mjsax I did, last night, but it is for a file (ByteArrayConverter on Connect) that I did not touch.

I think this is not from your code, but another PR merged recently that has a header issue, and it has been fixed in a most recent commit: https://github.com/apache/kafka/commit/962c378cc6cdb178541155af419c24e463516a5d

**Comment by nixsticks** [comment]
@mjsax @guozhangwang Rebased

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by nixsticks** [comment]
BTW, what is the generally followed process for squashing/rebasing commits before merging?

**Comment by mjsax** [comment]
We don't have an official policy. But this are my two cents:

 - As long as you work on the PR, it's best to just append new commits. This simplifies review process as the reviewer can inspect the deltas
 - If rebasing is required (eg., github shows some merge conflicts), just rebase (if possible don't squash commits -- but of course, if you squash first and rebase then, rebasing is simpler and it's also ok if you do this)

When a PR get's merged, the committer doing the merge will squash all commits anyway (so you don't need to squash by yourself). The commit message will contain the github title and github description for the PR (IRRC).

So overall, you don't need to worry about it too much :)

**Comment by ijuma** [comment]
The details can be found in the following wiki page:

https://cwiki.apache.org/confluence/display/KAFKA/Contributing+Code+Changes

What @mjsax is mostly right (apart from the `no official policy` comment :)). I'd just `git merge` instead of `git rebase` when you intend to update the branch with changes from trunk. Since commits are squashed automatically, there's not much point in using `rebase` and it doesn't make clear when the branch was updated to include more trunk commits.

**Comment by ijuma** [comment]
By the way, one issue I raised early on is that doing this client-side won't necessarily work in every case. Here's a concrete example:

https://issues.apache.org/jira/browse/KAFKA-4893

Probably still worthwhile doing it, but we still need to handle errors by the broker if it's stricter for whatever reason.

**Comment by mjsax** [comment]
@ijuma Thanks for clarification! :)

About KAFKA-4893 -- I guess, we should add a check to Streams, but this would be a different PR. The store names checked here are only part of topic names, and thus, we cannot check the topic name length at this level. I created https://issues.apache.org/jira/browse/KAFKA-4912 for this.

**Review by guozhangwang — APPROVED** [review]
Merged to trunk.

**Pull request closed without merging**
